### PR TITLE
[Platform API][PSU][Component] Compare applicable data against values from platform.json

### PR DESCRIPTION
**Pull request by sujinmkang**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### CODE REVIEW

File: tests/platform_tests/api/test_component.py [hunk 1]
```diff
@@ -48,6 +58,25 @@ def setup(self, platform_api_conn):
...
+    def compare_value_with_platform_facts(self, key, value, component_idx):
+        expected_value = None
+
+        if self.chassis_facts:
+            expected_components = self.chassis_facts.get("components")
+            if expected_components:
+                expected_value = expected_components[component_idx].get(key)
+
+        if not expected_value:
+            logger.warning("Unable to get expected value for '{}' from platform.json file for COMPONENT {}".format(key, component_idx))
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:74]: s/COMPONENT/component/ (no need for all caps here)

File: tests/platform_tests/api/test_component.py [hunk 2]
```diff
@@ -48,6 +58,25 @@ def setup(self, platform_api_conn):
...
+            return
+
+        self.expect(value == expected_value,
+                      "'{}' value is incorrect. Got '{}', expected '{}' for COMPONENT {}".format(key, value, expected_value, component_idx))
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:78]: s/COMPONENT/component/ (no need for all caps here)

File: tests/platform_tests/api/test_psu.py [hunk 3]
```diff
@@ -54,6 +82,7 @@ def test_get_name(self, duthost, localhost, platform_api_conn):
             name = psu.get_name(platform_api_conn, i)
             if self.expect(name is not None, "Unable to retrieve PSU {} name".format(i)):
                 self.expect(isinstance(name, STRING_TYPE), "PSU {} name appears incorrect".format(i))
+                compare_value_with_platform_facts(self, 'name', name, i)
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:85]: I believe you need to prefix with `self.`, as this is a class method.

File: tests/platform_tests/api/test_component.py [hunk 4]
```diff
@@ -60,6 +89,7 @@ def test_get_name(self, duthost, localhost, platform_api_conn):
             name = component.get_name(platform_api_conn, i)
             if self.expect(name is not None, "Component {}: Unable to retrieve name".format(i)):
                 self.expect(isinstance(name, STRING_TYPE), "Component {}: Name appears incorrect".format(i))
+                compare_value_with_platform_facts(self, 'name', name, i)
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:92]: I believe you need to prefix with `self.`, as this is a class method.

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by yxieca — COMMENTED** [review]
My only concern is that these warnings will be ignored. Had a discussion with Joe offline regarding this on another PR.

I think we could add a specific test to test that platform.json exists and that test can fail and indicating work needed. If the test could further validate required fields and failing on them too, then these warnings are fine.

**Comment by jleveque** [comment]
@yxieca: I am now of the belief that we should simply consider the platform.json file required, and fail any tests which rely on its data if it's not available. This will result in more consistency across platforms.

Here is a PR for the chassis tests: https://github.com/Azure/sonic-mgmt/pull/2268

File: tests/platform_tests/api/test_component.py [hunk 5]
```diff
@@ -32,10 +32,20 @@
     "next"
 ]
 
+@pytest.fixture(scope="class")
+def gather_facts(request, duthost):
+    # Get platform facts from platform.json file
+    request.cls.chassis_facts = duthost.facts.get("chassis")
+    if not request.cls.chassis_facts:
+        logger.warning("Unable to get chassis_facts from platform.json, test results will not be comprehensive")
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:40]: Remove warning message and make the check mandatory. See SFP portion of https://github.com/Azure/sonic-mgmt/pull/2268 for examples.

File: tests/platform_tests/api/test_component.py [hunk 6]
```diff
@@ -48,6 +58,25 @@ def setup(self, platform_api_conn):
...
+        expected_value = None
+
+        if self.chassis_facts:
+            expected_components = self.chassis_facts.get("components")
+            if expected_components:
+                expected_value = expected_components[component_idx].get(key)
+
+        if not expected_value:
+            logger.warning("Unable to get expected value for '{}' from platform.json file for component {}".format(key, component_idx))
+            return
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:75]: Remove warning message and make the check mandatory. See SFP portion of https://github.com/Azure/sonic-mgmt/pull/2268 for examples.

File: tests/platform_tests/api/test_psu.py [hunk 7]
```diff
@@ -32,10 +32,19 @@
 STATUS_LED_COLOR_RED = "red"
 STATUS_LED_COLOR_OFF = "off"
 
+@pytest.fixture(scope="class")
+def gather_facts(request, duthost):
+    # Get platform facts from platform.json file
+    request.cls.chassis_facts = duthost.facts.get("chassis")
+    if not request.cls.chassis_facts:
+        logger.warning("Unable to get chassis_facts from platform.json, test results will not be comprehensive")
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:40]: Remove warning message and make the check mandatory. See SFP portion of https://github.com/Azure/sonic-mgmt/pull/2268 for examples.

File: tests/platform_tests/api/test_psu.py [hunk 8]
```diff
@@ -45,6 +54,25 @@ def setup(self, platform_api_conn):
...
+        expected_value = None
+
+        if self.chassis_facts:
+            expected_psus = self.chassis_facts.get("psus")
+            if expected_psus:
+                expected_value = expected_psus[psu_idx].get(key)
+
+        if not expected_value:
+            logger.warning("Unable to get expected value for '{}' from platform.json file for PSU {}".format(key, psu_idx))
+            return
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:71]: Remove warning message and make the check mandatory. See SFP portion of https://github.com/Azure/sonic-mgmt/pull/2268 for examples.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/platform_tests/api/test_component.py [hunk 9]
```diff
@@ -70,11 +68,10 @@ def compare_value_with_platform_facts(self, key, value, component_idx):
             if expected_components:
                 expected_value = expected_components[component_idx].get(key)
 
-        if not expected_value:
-            logger.warning("Unable to get expected value for '{}' from platform.json file for component {}".format(key, component_idx))
-            return
+        pytest_assert(expected_value is not None,
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:71]: You should use a nested `self.expect()` here, as this gets called on a number of devices in the list, this way we can gather all the error messages. Please refer ot my SFP changes in https://github.com/Azure/sonic-mgmt/pull/2268, not the chassis changes.

File: tests/platform_tests/api/test_component.py [hunk 10]
```diff
@@ -70,11 +68,10 @@ def compare_value_with_platform_facts(self, key, value, component_idx):
...
+                      "Unable to get expected value for '{}' from platform.json file for component {}".format(key, component_idx))
 
-        self.expect(value == expected_value,
+        pytest_assert(value == expected_value,
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:74]: You should use a nested `self.expect()` here, as this gets called on a number of devices in the list, this way we can gather all the error messages. Please refer ot my SFP changes in https://github.com/Azure/sonic-mgmt/pull/2268, not the chassis changes.

File: tests/platform_tests/api/test_psu.py [hunk 11]
```diff
@@ -66,11 +65,10 @@ def compare_value_with_platform_facts(self, key, value, psu_idx):
             if expected_psus:
                 expected_value = expected_psus[psu_idx].get(key)
 
-        if not expected_value:
-            logger.warning("Unable to get expected value for '{}' from platform.json file for PSU {}".format(key, psu_idx))
-            return
+        pytest_assert(expected_value is not None,
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:68]: You should use `self.expect()` here, as this gets called on a number of devices in the list, this way we can gather all the error messages. Please refer ot my SFP changes in https://github.com/Azure/sonic-mgmt/pull/2268, not the chassis changes.

File: tests/platform_tests/api/test_psu.py [hunk 12]
```diff
@@ -66,11 +65,10 @@ def compare_value_with_platform_facts(self, key, value, psu_idx):
...
+                      "Unable to get expected value for '{}' from platform.json file for PSU {}".format(key, psu_idx))
 
-        self.expect(value == expected_value,
+        pytest_assert(value == expected_value,
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:71]: You should use a nested `self.expect()` here, as this gets called on a number of devices in the list, this way we can gather all the error messages. Please refer ot my SFP changes in https://github.com/Azure/sonic-mgmt/pull/2268, not the chassis changes.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/platform_tests/api/test_component.py [hunk 13]
```diff
@@ -48,6 +56,23 @@ def setup(self, platform_api_conn):
...
+    def compare_value_with_platform_facts(self, key, value, component_idx):
+        expected_value = None
+
+        if self.chassis_facts:
+            expected_components = self.chassis_facts.get("components")
+            if expected_components:
+                expected_value = expected_components[component_idx].get(key)
+
+        if self.expect(expected_value is not None,
+                      "Unable to get expected value for '{}' from platform.json file for component {}".format(key, component_idx))
```
Comment (jleveque) [inline tests/platform_tests/api/test_component.py:72]: This is now an `if` statement. It must end with a colon (`:`)

File: tests/platform_tests/api/test_psu.py [hunk 14]
```diff
@@ -45,6 +53,23 @@ def setup(self, platform_api_conn):
...
+    def compare_value_with_platform_facts(self, key, value, psu_idx):
+        expected_value = None
+
+        if self.chassis_facts:
+            expected_psus = self.chassis_facts.get("psus")
+            if expected_psus:
+                expected_value = expected_psus[psu_idx].get(key)
+
+        if self.expect(expected_value is not None,
+                      "Unable to get expected value for '{}' from platform.json file for PSU {}".format(key, psu_idx))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:69]: This is now an `if` statement. It must end with a colon (`:`)

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]

**Pull request merged**